### PR TITLE
Deprecate primes in identifiers exported from foreign modules

### PR DESCRIPTION
**Pull request by kl0tl**
This PR implements @hdgarrood [suggestion](https://github.com/purescript/purescript/issues/3613#issuecomment-583794497) regarding the depreciation of primes in foreign modules exports.

### CODE REVIEW

File: src/Language/PureScript/CodeGen/JS.hs [hunk 1]
```diff
@@ -312,7 +312,19 @@ moduleToJs (Module _ coms mn _ imps exps foreigns decls) foreign_ =
...
+        (AST.Ternary Nothing
+          (AST.App Nothing
+            (AST.Indexer Nothing (AST.StringLiteral Nothing "hasOwnProperty")
+                                 (AST.Var Nothing "$foreign"))
+            [AST.StringLiteral Nothing $ mkString unescaped])
+          (AST.StringLiteral Nothing $ mkString unescaped)
+          (AST.StringLiteral Nothing $ mkString (T.replace "'" "$prime" unescaped)))
+        (AST.Var Nothing "$foreign"))
+    | otherwise
+    = accessorString (mkString unescaped) (AST.Var Nothing "$foreign")
```
Comment (natefaubion) [inline src/Language/PureScript/CodeGen/JS.hs]: Is this emitted for every reference to a foreign identifier within a module?

**Comment by natefaubion** [comment]
Personally, I'm inclined to remove the `hasOwnProperty` check and just disallow foreign primes entirely rather than suggest you use `$prime` escaping.

**Comment by hdgarrood** [comment]
This looks like a great start, thank you! I think it might be preferable to use the existing `IfElse` rather than using a new `Ternary` constructor though (it slipped my mind that we don't already have one).

**Comment by hdgarrood** [comment]
> Personally, I'm inclined to remove the hasOwnProperty check and just disallow foreign primes entirely rather than suggest you use $prime escaping.

So I guess we would spend a major release series warning on FFI primes, and where the way of silencing the warning is to rename to something which doesn't contain primes at all, and then make it an error when we do ES modules? I guess that's preferable in that it doesn't involve messing with the generated code or `purs bundle` at all.

**Comment by hdgarrood** [comment]
Also it means that people are less likely to accidentally break their libraries on older compiler versions when dealing with the warning, which is probably also a plus.

File: src/Language/PureScript/CodeGen/JS.hs (hunk 1 above)
Comment (kl0tl) [reply src/Language/PureScript/CodeGen/JS.hs]: `$foreign[$foreign.hasForeignProperty("foo'") ? "foo'" : "foo$prime"]` would have been emitted for every reference to a `foo'` identifier imported from a foreign module yes.

**Comment by kl0tl** [comment]
I’ve removed the support for escaped primes and updated the deprecation message.

File: tests/purs/warning/DeprecatedFFIPrime.out [hunk 2]
```diff
@@ -0,0 +1,56 @@
+Error 1 of 4:
+
+  at tests/purs/warning/DeprecatedFFIPrime.purs:5:1 - 10:28 (line 5, column 1 - line 10, column 28)
+
+    In the FFI module for [33mMain[0m:
+
+      The identifier [33ma'[0m contains a [33m'[0m.
```
Comment (hdgarrood) [inline tests/purs/warning/DeprecatedFFIPrime.out]: I think people might not realise that a `'` is the same thing as a prime, which might make this wording confusing. How about having the error say this:

```
The identifier a' contains a prime (').
Primes in identifiers ...
```

Comment (kl0tl) [reply tests/purs/warning/DeprecatedFFIPrime.out]: I applied your suggestion!

**Comment by kl0tl** [comment]
Sorry for missing that, I just did.

**Comment by hdgarrood** [comment]
No worries - thanks very much!

**Comment by kl0tl** [comment]
I opened a few pull requests removing primes from foreign modules on the libraries I had to patch to be able to run tests while working on https://github.com/purescript/purescript/pull/3791. There’s probably a lot more libraries to upgrade in the official package set but that’s a start 😅

**Comment by garyb** [comment]
Thanks! I'll go through and get those all merged before we make the new release.

**Comment by kl0tl** [comment]
I rebased the PR onto master (`SimpleErrorMessage` moved from `Language.PureScript.AST.Declarations` to `Language.PureScript.Errors`) and opened pull requests on all other packages impacted by the depreciation in the set psc-0.13.6-20200309!

**Comment by hdgarrood** [comment]
Any objections to merging this now?

**Comment by natefaubion** [comment]
No objections from me.

**Pull request merged**